File: practices/chapter13_graph/python/ch_13_depth_first_search.py
def DFS(graph={}, start='', goal=''):
  solved_paths = []
  stack = [start]
  solved_res = False
  
  while stack:
    node = stack.pop()
    solved_paths.append(node)
    if node == goal:
      solved_res = True
      break
    for child in graph[node]:
      if child not in solved_paths:
        stack.append(child)

  return solved_paths

def recursive_dfs(graph, node, goal, solved_paths=[]):
  # print(node)
  solved_paths += [node]
  if node == goal:
    return solved_paths
  for child in graph[node]:
    if child not in solved_paths:
      solved_paths = recursive_dfs(graph, child, goal, solved_paths)
  return solved_paths

graph = {
  'A': ['B', 'C', 'D'],
  'B': ['A', 'E'],
  'C': ['A', 'F'],
  'D': ['A', 'G', 'H'],
  'E': ['B'],
  'F': ['C', 'I', 'J'],
  'G': ['D'],
  'H': ['D'],
  'I': ['F'],
  'J': ['F']
}

def main():
  solved = recursive_dfs(graph, 'F', 'G')
  # 兩者結果不同是正常的，因為一個是從後面(pop)取值，另一個是從頭取值
  print(solved)
  print(DFS(graph, 'F', 'G'))

if __name__ == "__main__":
    main()
    

File: practices/chapter13_graph/js/ch_13_depth_first_search.js
function DFS(graph, start, goal) {
  let findResult = false
  let visited = []

  if(!graph[start] || !graph[goal]) {
    return ({
      visited,
      findResult,
    })
  }

  let queue = [start]
  while(queue.length > 0) {
    // 跟BFS唯一的差別只在於從後面抓出元素(後進先出)，只要有新元素都先抓"最後面的"，所以是深度搜尋
    const node = queue.pop()

    if(node === goal) {
      visited.push(node)
      findResult = true
      break
    }

    if(!visited.includes(node)) {
      visited.push(node)
      queue.push(...graph[node])
    }
  }

  return ({
    visited,
    findResult,
  })
}


const dataGraph = {
  'A': ['B', 'C', 'D'],
  'B': ['A', 'E'],
  'C': ['A', 'F'],
  'D': ['A', 'G', 'H'],
  'E': ['B'],
  'F': ['C', 'I', 'J'],
  'G': ['D'],
  'H': ['D'],
  'I': ['F'],
  'J': ['F']
}

function main() {
  console.log('Result found: ', DFS(dataGraph, 'A', 'G'))
}

main()